docs/design/fabric-v2+/fpc-lifecycle-v2.puml

```diff
@@ -18,119 +15,126 @@ title Foot Box removed
+note over FPC_Peer1_CLI_Wrapper
+    Bookkeping operations.
+    Note: packageId is an input parameter of approveformyorg
 end note
-
-Admin_Org2 -> Orderer : Approve
-note left
-	**Chaincode definition**
-	Name: mycc, Version: 1
-	End Policy: (2of3), Seq 1,
-	Package ID: mycc_1:hash, <-???
-	//<color #00FF00>ECC_MR: mrenc_1</color>//
+FPC_Peer1_CLI_Wrapper -> FPC_Peer1_CLI_Wrapper : if packageId was stored then store cc_name, cc_version
+return /' Admin -> FPC_Peer1_CLI_Wrapper '/
+end /' group approve '/
+
+group other orgs approve
+Channel -> Channel : approvals from other parties
+note over Channel
+    Other participants must approve the chaincode definition,
+    and commit the corresponding transaction,
+    according to the LifecycleEndorsement policy.
+    So the party here has to loop to check for the approvals.
 end note
+end group 'other orgs approve'
 
-Admin_Org3 -> Orderer : Approve
-note left #FFAAAA
-	**Chaincode definition**
-	Name: mycc, Version: 1
-	End Policy: (2of3), Seq 1,
-	Package ID: mycc_1:hash, <-???
-	<color #FF0000>ECC_MR: mrenc_XXX</color>
-end note
 
 loop until enough approvals
-	Admin_Org1 -> Peer1_Org1 : checkcommitreadiness
+	Admin -> Peer1_CLI : lifecycle chaincode checkcommitreadiness
+note right
+    **Chaincode definition**
+    (implemented to match the approved one)
+end note
+    return
 end
 
-
-Admin_Org1 -> Peer1_Org1 : Commit proposal
-Admin_Org1 -> Peer1_Org2 : Commit proposal
-Admin_Org1 -> Peer1_Org3 : Commit proposal
-
-Peer1_Org1 --> Admin_Org1: proposal response
-Peer1_Org2 --> Admin_Org1: proposal response
-Peer1_Org3 --> Admin_Org1: proposal response
-
-Admin_Org1 -> Admin_Org1 : create transaction
-Admin_Org1 -> Orderer : submit commit transaction
-note left
-	**Chaincode definition**
-	Name: mycc, Version: 1
-	End Policy: (2of3), Seq 1,
-	Package ID: mycc_1:hash, <-???
-	//<color #00FF00>ECC_MR: mrenc_1</color>//
+group commit
+Admin -> FPC_Peer1_CLI_Wrapper ++:lifecycle chaincode commit
+FPC_Peer1_CLI_Wrapper -> Peer1_CLI ++: lifecycle chaincode commit
+note right
+    **Chaincode definition**
+    (implemented to match the one that is ready to be committed)
 end note
-
-
-@enduml
\ No newline at end of file
+Peer1_CLI -> Orderer ++: commit transaction
+return transaction committed /' Peer1_CLI -> Orderer '/
+return /' FPC_Peer1_CLI_Wrapper -> Peer1_CLI '/
+note over FPC_Peer1_CLI_Wrapper
+    Commit has succeeded.
+    Now, check if this fpc cc was previously approved.
+    If so, perform some invocations.
+end note
+FPC_Peer1_CLI_Wrapper -> FPC_Peer1_CLI_Wrapper : if cc_name, cc_version not stored then return
+FPC_Peer1_CLI_Wrapper -> Peer1_CLI ++: invoke chaincode __setup
+Peer1_CLI -> Orderer ++: invoke transaction
+return transaction committed
+return /' FPC_Peer1_CLI_Wrapper -> Peer1_CLI '/
+FPC_Peer1_CLI_Wrapper -> Peer1_CLI ++: query chaincode __getEnclavePk
+return /' FPC_Peer1_CLI_Wrapper -> Peer1_CLI '/
+return /' Admin -> FPC_Peer1_CLI_Wrapper '/
+end /' group commit '/
+
```

Comment:
The `__setup` is not consistent with name in the registration protocol flow (this was also an observation by Marcus in the original review). It is also not clear whether we really want to hide all of that behind `commit`? It works for designated peer but definitely not beyond for 
 single peer or multi-peer endorsement? So for future-proofing seems better to have a separate admin command?.  This really also depends from what we decided to on first defining the peer cli management interface as part of #358 in a separate markdown rather than defining this interface in the UML as initially thought. (see, though, why below proposal elegantly side-steps that for this diagram :-) Won't work for the registration diagram though ...)  

Also `__getEnclavePK` is superfluous (even though i realize that it exists i think still in the actual cli wrapper).

Proposal would be to drop completely and just reference here that registration and key-managements follow with just a reference to these diagrams. Below a suggestion (but note i haven't tested whether it looks nice in PUML!)

```suggestion
end /' group commit '/
note left
  To complete setting up an FPC chaincode, enclaves have to be registered and provisioned with keys.  See the separate UML diagrams `fpc-registration` and `fpc-key-dist` for the corresponding admin commands and protocol flows.
end note
```

Reply:
07b4ce2a989f777ea4f0b16722ca2a6274ce6a15

Reply:
I have two concerns here going forward:
1) we have a diagram that does not match the code
2) most importantly, given the confusion, we should consider
- dropping the wrapper (for the short-term)
- implementing the commands along side Fabric's ones (just as Jason suggested)
- once FPC is mature, then consolidate FPC and Fabric commands (possibly in a wrapper)

Reply:
Regarding 1., well, we haven't changed anything in the code yet, so how can the new design match the code?  Inevitably, the design is always ahead of the code?!  Regarding the wrapper, i don't see how that is contrary to what Jason said: we are exposing additional user actions as Jason suggested but we do it in a way which is future proof.  In any case, i thought we decided already in the registration diagram discussion to handle user (admin) actions in the cli component (which essentially is the cli+wrapper)?

Reply:
Regarding 2), in the lifecycle diagram I would drop the differentiation between cli and cli wrapper and just call it `peer cli` as used by the admin to talk to a peer. As mentioned earlier the wrapper is just our way to add additional functionality to the fabric peer cli, an alternative would be to indeed add our functionality to the cli code and recompile. 

As this diagram should illustrate the conceptional changes we need for FPC to work. 
Let me ask the question what are the major changes here? 

For lifecycle ...

- do we add any additional management commands needed for FPC? - No
- what are the crucial steps?

1. package and install: FPC chaincode comes as external builder package

1. approveformyorg : Here FPC changes are important as we need to set MRENCLAVE and ecc-vscc, this is the way all orgs agree on MRENCLAVE

Regarding the details on storing packageID inside the CLI (wrapper), I understand this is a nice helper to abstract away FPC complexity but I see this as an implementation specific decision to let our CLI implementation autofill the chaincode version with MRENCLAVE and validation plugin with ecc-vscc. I am questioning if this is relevant for the purpose of this document.

Reply:
@mbrandenburger  Ah, now, i see that you actually did add an explanation to your change below :-) [and i think it is a good suggestion: the cli component is essentially our management API, many of the commands as-is from fabric, some with slightly refined semantics (approveformyorg/commit require version to be mrenclave) and some new (enclave creation & registration etc)